### ISSUE TITLE
Support linter config files in reusable workflow repos

### ISSUE THREAD

**Issue by jasonkarns** [super-linter/super-linter#7984]
### Is there an existing issue for this?

- [x] I have searched the existing issues

### Current Behavior

The codespell linter is raising false positives in man-page files. These files contain all kinds of syntax that codespell can't currently parse correctly.

The super-linter configuration (out of the box) is making codespell attempt to parse these files.

### Expected Behavior

I would expect man-files to be ignored for the same reason that the current codespell template ignores other generated files like package-lock.json.

**Comment by ferrarimarco** [comment]
Hi @jasonkarns, thanks for raising this issue.

Can you please give me an example of that?

**Comment by jasonkarns** [comment]
Here's a test run that is failing due to codespell hitting man pages: https://github.com/nodenv/nodenv/actions/runs/28947599090

<img width="867" height="231" alt="Image" src="https://github.com/user-attachments/assets/cd685208-e565-4c31-bee6-b719f2afa9ef" />

**Comment by ferrarimarco** [comment]
Thanks for the example. My understanding of the [Codespell docs about ignoring stuff](https://github.com/super-linter/super-linter/blob/main/TEMPLATES/.codespellrc) is that the `skip` directive supports excluding files by their filenames, and supports regular expressions.

This is how super-linter configures Codespell ([link to Codespell configuration file (.codespellrc)](https://github.com/super-linter/super-linter/blob/main/TEMPLATES/.codespellrc)):

https://github.com/super-linter/super-linter/blob/07267f14343a39804854682b9e28c3bacb514e05/TEMPLATES/.codespellrc#L1-L3

I am not sure that we can find a general path pattern to exclude man files.

My suggestion is to have your own `.codespellrc`, ignoring files in your `share/man/man1` directory.

**Comment by jasonkarns** [comment]
The challenge I'm trying to navigate is with reusable workflows.

super-linter is driven by the workflow in [`nodenv/.github`](https://github.com/nodenv/.github). With all the linted repos being `nodenv/*`. If the only way to control codespell is with files in the individual repos, that completely negates the value of having a reusable workflow, because the codespellrc needs to be duplicated across all the repos.

And the duplication doesn't even address the "spooky action at a distance" anti-pattern that would exhibit. Having the codespellrc file driving codespell configuration live in the repo, is now "very far" from the codespell linting _tool_ which is configured in the reusable workflow repo (nodenv/.github). And this isn't an issue limited to just codespell, it's a symptom of the larger issue with superlinter (arguably a reflection of limitations with reusable workflows in general).

Most other tools avoid this problem by having configurability through env vars, which _can_ be set at the reusable workflow level (and thus shared). But superlinter might be able to solve this by introducing a means to have these rc files live in/with the reusable workflow and mounted in a way that doesn't require them be copied to all the calling workflow repos.

**Comment by ferrarimarco** [comment]
Thanks for sharing additional context. This is not just an issue with Codespell, as you say, but also with any tool that's configurable using a file.

My understanding of your request is that you'd like a way to configure linters/formatters using environment variables, so that you can bake your super-linter configuration in the workflow step that calls the super-linter reusable workflow.

As you point out, you're hitting a limitation of how reusable workflows are implemented. I think there's room to implement something that allows for configuring tools using environment variables, but that would require some rewriting of the logic that loads and validates configuration files.

If my understanding is correct, I think we should rename this feature request to reflect the general "configurability through environment variables" angle, rather than focusing on Codespell alone.

I'm not sure there are practical work arounds, besides having a step in your reusable workflow that downloads configuration files from a source of truth, and puts them where Super-linter expects them for discovery.

**Comment by jasonkarns** [comment]
> My understanding of your request is that you'd like a way to configure linters/formatters using environment variables, so that you can bake your super-linter configuration in the workflow step that calls the super-linter reusable workflow.

Yes, perhaps. But I think that's only one option. The end goal is for the linter configuration to live _with_ the super-linter workflow. The super-linter invocation lives in a reusable workflow. One option would be to have configuration defined via env vars that can be set in that workflow yml. Another option would be for super-linter to include files that live in that workflow's repo.

contrived example:

reusable workflows repo: `someorg/workflows`
reusable super-linter workflow: `someorg/workflows/.github/workflows/lint.yml`
linter configs (ie, `LINTER_RULES_PATH`): `someorg/workflows/.github/linters`

**Comment by ferrarimarco** [comment]
The configurability using environment variables is something that can be scoped to be a supported feature, I think.

Downloading arbitrary configuration files seems a bit out of scope tho? Wouldn't it be better suited for the reusable workflow implementation, as in the workaround I described?

In your example, the reusable workflow could have a "configuration init" step to download those configuration files from `someorg/workflows/.github/linters`.

**Comment by jasonkarns** [comment]
> Downloading arbitrary configuration files seems a bit out of scope tho?

Yes, I'd say so. That could be accomplished trivially by the reusable workflow itself. Github already provides context variables that reference the workflow repo and sha, so the reusable workflow repo can be checked out explicitly, and at the same version as the one "currently executing".

The documentation for super-linter doesn't seem to indicate whether `LINTER_RULES_PATH` is a true path variable (that is, colon-separated multi-value). If it is, then a workaround would be for the reusable workflow to do an explicit checkout of the reusable workflow repo, and set the `LINTER_RULES_PATH` to cover both directories:

`LINTER_RULES_PATH=.github/linters:somedir/.github/linters`

I think it's more the purview of super-linter so handle resolution of config files from multiple directories, than to deal with downloading additional files from arbitrary locations. Indeed, super-linter already handles multiple config file resolution (it looks in .github/linters first, but has a fallback to the baked in templates). Handling proper colon-separated-path would allow for reusable workflow configs to be included in that lookup without much hassle.

(And it could begin a cowpath whereby github establishes a convention for reusable workflow repos themselves to be cloned into a consistent location. say to `.github/workflows` itself. Like `.github/workflows/ORG/REPO`.)

I think I'm narrowing in on `LINTER_RULES_PATH` being a proper multi-value path, as the feature request here. (Unless it already supports colon-separated values.) This would ensure that some the reusable workflow repo could provide "default" linter rule configs for some tools, while the some calling repos could provide any linter rule configs that need overridden.

**Comment by jasonkarns** [comment]
Not-completely-irrelevant to this thread, but I'm suddenly more annoyed than usual about the lack of adoption of XDG basedir conventions. (by various tools)

In an alternate reality where all these linter tools adhered to the XDG basedir spec, then super-linter wouldn't need to do _anything_. (And `LINTER_RULES_PATH` wouldn't even need to exist.)

One could just append additional directories to `XDG_CONFIG_DIRS`. (Imagine calling-workflows that added `.github/linters` to XDG_CONFIG_DIRS. And reusable workflows cloning themselves into `.github/workflows/USER/REPO` while appending `.github/workflows/USER/REPO/.github/linters` to XDG_CONFIG_DIRS. The whole thing would Just Work and super-linter wouldn't need to even be aware.

**Comment by jasonkarns** [comment]
Attempting a half workaround...

The reusable workflow now clones itself into the workspace:

https://github.com/nodenv/.github/blob/51d99cab38ca42ec327ba45e596f8815abb961de/.github/workflows/test.yml#L42-L47

```
# eg: .github/workflows/nodenv/.github
path: .github/workflows/${{ job.workflow_repository }}
```

Thus allowing the reusable workflow to reference linter config files that exist in the reusable workflow's repo (at `.github/linters`).

https://github.com/nodenv/.github/pull/166

```yml
# eg .github/workflows/nodenv/.github/.github/linters/.codespellrc
SPELL_CODESPELL_CONFIG_FILE: .github/workflows/${{ job.workflow_repository }}/.github/linters/.codespellrc
```

codespellrc is definitely there:

<img width="973" height="110" alt="Image" src="https://github.com/user-attachments/assets/315b523d-76e4-4f98-89d2-b1a1d264ca5d" />

But super-linter doesn't think so:

<img width="1600" height="79" alt="Image" src="https://github.com/user-attachments/assets/48f98ed1-12c5-42f8-b281-345bdc14dccf" />

Is super-linter's current working directory different somehow? I'm not clear on super-linter can't seem to find the file. The working directory is the same default that github actions sets. (Not explicitly set or changed by the workflow in any way.)

Full run here: https://github.com/nodenv/.github/actions/runs/29952180548/job/89032320710

(failing on both push and pull_request triggers. Though 'push' is a bit easier to follow because the manual checkout of the reusable repo isn't pointing to a MERGE_HEAD commit.)


edit: updated to show the current working directory when listing the codespellrc file: 

<img width="982" height="131" alt="Image" src="https://github.com/user-attachments/assets/14f24c67-66d0-422d-b94a-7e207b5db8a8" />

Why is superlinter's `SPELL_CODESPELL_CONFIG_FILE` operating from a different working directory? (`/action/lib/.automation/` instead of `/home/runner/work/.github/.github`)

**Comment by jasonkarns** [comment]
Specifying an absolute path doesn't work either:

```
SPELL_CODESPELL_CONFIG_FILE: ${{ github.workspace }}/.github/workflows/${{ job.workflow_repository }}/.github/linters/.codespellrc
```

Still prepends: `/action/lib/.automation/` (which also results in double-slash because the paths aren't being normalized: `/action/lib/.automation//github/workspace/....`)

<img width="1668" height="76" alt="Image" src="https://github.com/user-attachments/assets/8502428e-03ff-4093-999b-50b3a473cde4" />

**Comment by ferrarimarco** [comment]
Those `_CONFIG_FILE` variables are for specifying the config file **names**, which are relative to `LINTER_RULES_PATH`. I'd suggest setting `LINTER_RULES_PATH` to where you put the config files you download / clone when running the reusable workflow.

`LINTER_RULES_PATH`, in turns, is relative to `DEFAULT_WORKSPACE`. When running on GitHub Actions, `DEFAULT_WORKSPACE` is set to `GITHUB_WORKSPACE`.

The implication is that you should make configuration files available in `GITHUB_WORKSPACE`, and set `LINTER_RULES_PATH` accordingly. For example, you can download / clone configuration files in `${GITHUB_WORKSPACE}/global-super-linter-config`, or something like that, and set `LINTER_RULES_PATH` to `global-super-linter-config`.

Some pointers from the README that might be helpful:

- https://github.com/super-linter/super-linter#configure-linters
- See the `DEFAULT_WORKSPACE` definition in the https://github.com/super-linter/super-linter#configure-super-linter section.

**Comment by jasonkarns** [comment]
Thanks for clarifying.

So to make sure I understand this, with a design like this, it means it's impossible for users to have a config file for tool-A in a different directory than tool-B?

Followup question: if I do as you suggest to set `LINTER_RULES_PATH` to a directory where the cloned reusable repo puts its config files... then that would mean the calling repo would not be able to provide any config files of its own in its `.github/linters` path, right?

So the limitation that I'm clarifying above (super-linter can only handle a single directory of config files for _all tools_), means that I can't point, say, codespell to a "shared" config file (provided by a clone of the reusable workflow repo), while still allowing, say, eslint to use the config file in the calling repo's `.github/linters` directory.


(Unrelated to the question above: there might be some room to improve the docs. Currently the readme only has 2 instances of the word "relative", and neither indicate that each tool's config file variable _in general_ are all relative to the LINTER_RULES_PATH. `OS_PACKAGES_CONFIG_FILE_NAME` does explicitly mention its file name is relative to LINTER_RULES_PATH. But `SPELL_CODESPELL_CONFIG_FILE` does not. And I don't see any mention in general that indicates the config file vars are _all_ relative to the LINTER_RULES_PATH.
